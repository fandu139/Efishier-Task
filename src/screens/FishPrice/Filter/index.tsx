import React from 'react';
import { View, StyleSheet, Button, Text, FlatList, TouchableOpacity } from 'react-native';
import Colors from '../../../theme/colors';
import Fonts from '../../../theme/fonts';
import AppStyles from '../../../theme/appStyles';
import Checkbox from '../../../uikit/Checkbox';
import Icon from '../../../uikit/Icon';
import { goBack } from '../../../helper/navigation';
import { ICON_ARROW_DOWN, ICON_ARROW_UP } from '../../../assets/icon';
import SteinStore from 'stein-js-client';
import { useEffect } from 'react';
import { useState } from 'react';
import { StackScreenProps } from '@react-navigation/stack';

type RouteParams = {
  handleFilter: (type: string, text: string) => void;
};

type RootStackParamList = {
  ListFilterScreen: RouteParams;
};

type Props = StackScreenProps<RootStackParamList, 'ListFilterScreen'>;

const storeArea = new SteinStore(
  'https://stein.efishery.com/v1/storages/5e1edf521073e315924ceab4/option_area',
);
const storeSize = new SteinStore(
  'https://stein.efishery.com/v1/storages/5e1edf521073e315924ceab4/option_size',
);

const ListFilterScreen: React.FC<Props> = ({ route }: Props) => {
  const { handleFilter } = route.params;
  const [categoryFilterSelected, setCategoryFilterSelected] = useState(0)
  const [dataSelected, setDataSelected] = useState({
    idCategoryFilterSelected: null,
    dataSelect: null
  });
  const listDataFilter = [{
    id: 0,
    label: 'Filter berdasarkan provinsi dan kota',
  }, {
    id: 1,
    label: 'Filter berdasarkan size',
  }];
  const [filterByArea, setFilterByArea] = React.useState([]);
  const [filterBySize, setFilterBySize] = React.useState([]);

  useEffect(() => {
    const getData = async () => {
      storeArea.read('').then(data => {
        setFilterByArea(data)
      });

      storeSize.read('').then(data => {
        setFilterBySize(data)
      });
    };

    getData();
  }, []);

  const clearFilter = () => {
    setDataSelected({
      idCategoryFilterSelected: null,
      dataSelect: null
    })
  }

  const handleFilterData = () => {
    const type = categoryFilterSelected ? 'size' : 'city'
    const data = categoryFilterSelected ? filterBySize : filterByArea;
    handleFilter(type, data[dataSelected?.dataSelect]);
    goBack();
  }

  const RenderSelectedItem = ({ item }) => {
    const data = item?.id ? filterBySize : filterByArea;

    return (
      <View style={styles.topContent}>
        {data.map((itemList, indexList) => {
          const testID = item?.id ? itemList?.size : `${itemList?.province} - ${itemList?.city}`;
          const label = item?.id ? `Size - ${itemList?.size}` : `${itemList?.province} - ${itemList?.city}`;

          return (
            <View key={indexList} style={styles.bottomContentContainer}>
              <View style={styles.buttonContainerWithBorder}>
                <Checkbox
                  testID={testID}
                  onChangeValue={() => {
                    setDataSelected({
                      idCategoryFilterSelected: item?.id,
                      dataSelect: indexList
                    });
                  }}
                  checked={dataSelected?.dataSelect === indexList &&  dataSelected?.idCategoryFilterSelected === item?.id }
                  label={label}
                />
              </View>
            </View>
          )
        })}
      </View>
    )
  }

  return (
    <View style={styles.svContentContainer}>
      <FlatList
        onEndReachedThreshold={0.1}
        style={styles.container}
        data={listDataFilter}
        keyExtractor={(_, index) => index.toString()}
        renderItem={({ item, index }) => {
          const icon = categoryFilterSelected === item.id ? ICON_ARROW_UP : ICON_ARROW_DOWN
          return (
            <View style={styles.listItemContainer}>
              <TouchableOpacity
                key={item.id} 
                onPress={() => {
                  if (index === categoryFilterSelected) {
                    setCategoryFilterSelected(null);
                  } else {
                    setCategoryFilterSelected(index);
                  }
                }}
                style={styles.listCategoryContainer}
              >
                <Text style={styles.titleLabelCheckbox}>{item.label}</Text>
                <Icon
                  testID="icon-check-available"
                  accessibilityLabel="icon-check-available"
                  name={icon}
                  size={10}
                  color={Colors.RED}
                />
              </TouchableOpacity>
              {item?.id === categoryFilterSelected && <RenderSelectedItem item={item} />}
            </View>
          );
        }}
      />
      <View style={styles.actionContainer}>
        <View style={styles.filterContainer}>
          <Button
            testID="button-filter-order"
            onPress={handleFilterData}
            title={'Filter'}
          />
        </View>
        <View>
          <Button
            testID="button-clear-filter"
            onPress={clearFilter}
            title={'Reset'}
            color={Colors.RED}
          /> 
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: 15,
  },
  buttonContainer: {
    marginVertical: 15,
  },
  svContentContainer: {
    flexGrow: 1,
    flexDirection: 'column',
    justifyContent: 'space-between',
    backgroundColor: Colors.WHITE,
  },
  bottomContentContainer: {
    marginVertical: 10,
    borderBottomWidth: 1,
    borderBottomColor: Colors.GRAYE0,
  },
  labelCheckbox: {
    marginLeft: 10,
  },
  topContent: {
    paddingHorizontal: 20,
  },
  buttonContainerWithBorder: {
    ...AppStyles.rowItemsCenter,
    paddingBottom: 20,
  },
  titleLabelCheckbox: {
    fontSize: Fonts.size.extraSmall,
    color: Colors.GRAY60,
    paddingVertical: 20,
    flexGrow: 1,
  },
  listItemContainer: {
    backgroundColor: Colors.WHITE,
    paddingHorizontal: 20,
    borderBottomWidth: 1,
    borderBottomColor: Colors.GRAYF3,
  },
  listCategoryContainer: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  actionContainer: {
    backgroundColor: Colors.WHITE,
    padding: 20,
    borderRadius: 5,
    borderTopWidth: 1,
    borderColor: Colors.GRAYEF,
  },
  filterContainer: {
    paddingBottom: 20,
  }
});

export default ListFilterScreen;
