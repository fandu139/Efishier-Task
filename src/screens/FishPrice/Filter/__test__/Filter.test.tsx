import Filter from '../index';
import React from 'react';
import { render } from '@testing-library/react-native';

describe('Filter screen test', () => {
  describe('Filter screen snapshot test', () => {
    it('should render correctly', () => {
      const navigation = {
        setOptions: jest.fn(),
        replace: jest.fn(),
      };

      const mockedRoute = {
        params: {
          handleFilter: () => null
        },
      };

      const tree = render(<Filter navigation={navigation as never} route={mockedRoute as never} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
