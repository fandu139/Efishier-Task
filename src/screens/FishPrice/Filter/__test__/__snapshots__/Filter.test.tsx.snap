// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter screen test Filter screen snapshot test should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flexDirection": "column",
      "flexGrow": 1,
      "justifyContent": "space-between",
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": 0,
          "label": "Filter berdasarkan provinsi dan kota",
        },
        Object {
          "id": 1,
          "label": "Filter berdasarkan size",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={0.1}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
        "paddingTop": 15,
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "borderBottomColor": "#F3F3F3",
              "borderBottomWidth": 1,
              "paddingHorizontal": 20,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#606060",
                  "flexGrow": 1,
                  "fontSize": 12,
                  "paddingVertical": 20,
                }
              }
            >
              Filter berdasarkan provinsi dan kota
            </Text>
            <View
              accessibilityLabel="icon-check-available"
              accessibilityState={
                Object {
                  "disabled": true,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={false}
              hitSlop={
                Object {
                  "bottom": 15,
                  "left": 15,
                  "right": 15,
                  "top": 15,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="icon-check-available"
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 10,
                    },
                    Object {
                      "color": "#EF4941",
                    },
                    Object {
                      "fontFamily": "ColectionIcon",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                юдК
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "paddingHorizontal": 20,
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "borderBottomColor": "#F3F3F3",
              "borderBottomWidth": 1,
              "paddingHorizontal": 20,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#606060",
                  "flexGrow": 1,
                  "fontSize": 12,
                  "paddingVertical": 20,
                }
              }
            >
              Filter berdasarkan size
            </Text>
            <View
              accessibilityLabel="icon-check-available"
              accessibilityState={
                Object {
                  "disabled": true,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={false}
              hitSlop={
                Object {
                  "bottom": 15,
                  "left": 15,
                  "right": 15,
                  "top": 15,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="icon-check-available"
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 10,
                    },
                    Object {
                      "color": "#EF4941",
                    },
                    Object {
                      "fontFamily": "ColectionIcon",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                юдв
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderColor": "#EFEFEF",
        "borderRadius": 5,
        "borderTopWidth": 1,
        "padding": 20,
      }
    }
  >
    <View
      style={
        Object {
          "paddingBottom": 20,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="button-filter-order"
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Filter
          </Text>
        </View>
      </View>
    </View>
    <View>
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="button-clear-filter"
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#EF4941",
                },
              ]
            }
          >
            Reset
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
